plugins/inputs/logstash/logstash.go

```diff
@@ -0,0 +1,290 @@
+		if err != nil {
+			return err
+		}
+		acc.AddFields("logstash_"+p, f.Fields, tags, now)
+	}
+
+	return nil
+}
+
+func (l *Logstash) gatherProcessStats(url string, acc telegraf.Accumulator) error {
+	ProcessStats := &ProcessStats{}
+	if err := l.gatherJSONData(url, ProcessStats); err != nil {
+		return err
+	}
+
+	tags := map[string]string{
+		"node_id": ProcessStats.CommonData.ID,
+	}
+
+	stats := map[string]interface{}{
+		"process": ProcessStats.Process,
+	}
+
+	now := time.Now()
+	for p, s := range stats {
+		f := jsonparser.JSONFlattener{}
+		// parse Json, ignoring strings and bools
+		err := f.FlattenJSON("", s)
+		if err != nil {
+			return err
+		}
+		acc.AddFields("logstash_"+p, f.Fields, tags, now)
+	}
+
+	return nil
+}
+
+func (l *Logstash) gatherPipelineStats(url string, acc telegraf.Accumulator) error {
+	PipelineStats := &PipelineStats{}
+	if err := l.gatherJSONData(url, PipelineStats); err != nil {
+		return err
+	}
+
+	tags := map[string]string{
+		"node_id": PipelineStats.CommonData.ID,
+	}
+
+	stats := map[string]interface{}{
+		"events": PipelineStats.Pipeline.Events,
+	}
+
+	now := time.Now()
+	// Events
+	for p, s := range stats {
+		f := jsonparser.JSONFlattener{}
+		// parse Json, ignoring strings and bools
+		err := f.FlattenJSON("", s)
+		if err != nil {
+			return err
+		}
+		acc.AddFields("logstash_"+p, f.Fields, tags, now)
+	}
+
+	// Input Plugins
+	for _, plugin := range PipelineStats.Pipeline.Plugins.Inputs {
+		//plugin := &plugin
+		fields := map[string]interface{}{
+			"name":               plugin.Name,
+			"duration_in_millis": plugin.Events.DurationInMillis,
+			"in":                 plugin.Events.In,
+			"out":                plugin.Events.Out,
+		}
+		acc.AddFields("logstash_plugin_input_"+plugin.Name, fields, tags, now)
+	}
+
+	// Filters Plugins
+	for _, plugin := range PipelineStats.Pipeline.Plugins.Filters {
+		//plugin := &plugin
+		fields := map[string]interface{}{
+			"name":               plugin.Name,
+			"duration_in_millis": plugin.Events.DurationInMillis,
+			"in":                 plugin.Events.In,
+			"out":                plugin.Events.Out,
+		}
+		acc.AddFields("logstash_plugin_filter_"+plugin.Name, fields, tags, now)
+	}
+
+	// Output Plugins
+	for _, plugin := range PipelineStats.Pipeline.Plugins.Outputs {
+		//plugin := &plugin
+		fields := map[string]interface{}{
+			"name":               plugin.Name,
+			"duration_in_millis": plugin.Events.DurationInMillis,
+			"in":                 plugin.Events.In,
+			"out":                plugin.Events.Out,
+		}
+		acc.AddFields("logstash_plugin_output_"+plugin.Name, fields, tags, now)
+	}
+
+	return nil
+}
+
+//Gather is main function to gather all metrics provided by this plugin
+func (l *Logstash) Gather(acc telegraf.Accumulator) error {
+
+	if l.client == nil {
+		client, err := l.createHTTPClient()
+
+		if err != nil {
+			return err
+		}
+		l.client = client
+	}
+
+	if err := l.gatherJVMStats(l.LogstashURL+jvmStats, acc); err != nil {
```

Comment:
Use the url library to build the urls, this will help prevent errors with malformed paths.